core/src/apps/cardano/helpers/tx_builder_state_machine.py

```diff
@@ -0,0 +1,158 @@
+from micropython import const
+
+if False:
+    ActionType = int
+
+
+class TxBuilderStateMachine:
+    # states
+    STATE_INIT = const(0)
+    STATE_INPUTS = const(1)
+    STATE_INPUTS_DONE = const(2)
+    STATE_OUTPUTS = const(3)
+    STATE_OUTPUT_WITH_TOKENS = const(4)
+    STATE_ASSET_GROUPS = const(5)
+    STATE_TOKENS = const(6)
+    STATE_ASSET_GROUPS_DONE = const(7)
```

Comment:
if the action is suffixed with finished, I'd consider suffixing the resulting states with "FINISHED" for the sake of consistency